SRVKE-1520: Add transport-encryption documentation (#107)
openshift-knative/docs: modules/serverless/pages/features/transport-encryption/transport-encryption-setup.adoc

= Setup transport encryption in Eventing
:compat-mode!:
// Metadata:
:description: Setup {serverlessproductname} transport encryption for Eventing

This page describes the transport encryption feature which allows transporting data and events over secured and encrypted HTTPS connections using TLS.

[IMPORTANT]
====
{serverlessproductname} transport encryption for Eventing is a Developer Preview feature only.
Developer Preview features are not supported with Red Hat production service level agreements (SLAs) and might not be functionally complete.
Red Hat does not recommend using them in production.
These features provide early access to upcoming product features, enabling customers to test functionality and provide feedback during the development process.

For more information about the support scope of Red Hat Developer Preview features, see https://access.redhat.com/support/offerings/devpreview/.
====

== Prerequisites

* You have access to an {product-title} account with cluster administrator access.

> * Install the OpenShift CLI (`oc`).
> 
> * Install the {serverlessoperatorname}.
> 
> * Install the {certmanageroperatorname}.

Reviewer: ```suggestion
* OpenShift CLI (`oc`) is installed.

* {serverlessoperatorname} is installed.

* {certmanageroperatorname} is installed.
```

nit. Maybe easier to read as it is a "prerequisite" :shrug:

Author: I copied it from here https://openshift-knative.github.io/docs/docs/latest/serverless/service-mesh/common-service-mesh-setup.html

![image](https://github.com/openshift-knative/docs/assets/33736985/2cdf45fb-6cc1-4f4c-be1c-b9b32d6dfa8a)

== Setup a `SelfSigned` `ClusterIssuer`

> > `ClusterIssuers`, are Kubernetes resources that represent certificate authorities (CAs) that are able to generate signed certificates by honoring certificate signing requests.

Reviewer 2: This seems rendered indented. Should this be a info block or something?

Author: It is supposed to be a citation but info block sounds good

All cert-manager certificates require a referenced issuer that is in a ready condition to attempt to honor the request. +
Reference: https://cert-manager.io/docs/concepts/issuer/

[IMPORTANT]
====

> For the simplicity of this guide, we will use a `SelfSigned` issuer as root certificate, however, be aware of the implications and limitations as documented at https://cert-manager.io/docs/configuration/selfsigned/ of this method.

Reviewer 2: Maybe also add that we do not recommend this for production use?

Author: what would be our recommendation? use the same PKI as the configured OpenShift PKI or something else ?

Reviewer 2: Highly depends on what they (the customers company) already have. For a smooth rotation, for production, at least we should recommend to use the CA issuer (where they provide their own CA as a Secret).

Author: Right, I think I will stay vague here and just link cert-manager docs

====

. Create a `SelfSigned` `ClusterIssuer`:
+
[source,yaml]
----
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: knative-eventing-selfsigned-issuer
spec:
  selfSigned: {}
----
+
. Apply the `ClusterIssuer` resource:
+
[source,terminal]
----
$ oc apply -f <filename>
----

. Create a root certificate using the previously created `SelfSigned` `ClusterIssuer`:
+
[source,yaml]
----
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: knative-eventing-selfsigned-ca
  namespace: cert-manager <1>
spec:
  secretName: knative-eventing-ca <2>

  isCA: true
  commonName: selfsigned-ca
  privateKey:
    algorithm: ECDSA
    size: 256

  issuerRef:
    name: knative-eventing-selfsigned-issuer
    kind: ClusterIssuer
    group: cert-manager.io

----
+
<1> The {certmanageroperatorname} namespace, `cert-manager` by default.

> <2> Secret name that will be later used for the `ClusterIssuer` for Eventing

Reviewer: ```suggestion
<2> Secret name later used for the `ClusterIssuer` for Eventing
```

+
. Apply the `Certificate` resource:
+
[source,terminal]
----
$ oc apply -f <filename>
----

== Setup `ClusterIssuer` for Eventing

. Create the `knative-eventing-ca-issuer` `ClusterIssuer` for Eventing:
+
[source,yaml]
----
# This is the issuer that every Eventing component should use to issue their server's certs.
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: knative-eventing-ca-issuer
spec:
  ca:
    secretName: knative-eventing-ca <1>
----
+
<1> Secret name in the {certmanageroperatorname} namespace (`cert-manager` by default) containing the certificate that can then be used by Knative Eventing components for new certificates.
+
[IMPORTANT]
====
The name of the `ClusterIssuer` must be `knative-eventing-ca-issuer`.
====
+
. Apply the `ClusterIssuer` resource:
+
[source,terminal]
----
$ oc apply -f <filename>
----

== Understanding the transport encryption configuration

The transport-encryption feature flag is an enum configuration that configures how Addressables (Broker, Channel, Sink) should or must accept events.

The possible values for `transport-encryption` are:

> * *disabled* (this is equivalent to the current behavior)

Reviewer: ```suggestion
* *`disabled`* (this is equivalent to the current behavior)
```
(only not sure if this is supported in adoc)

Reviewer 2: I wanted to suggest the same. I think is supported, as above we have `transport-encryption`

** Addressables may accept events to HTTPS endpoints
** Producers may send events to HTTPS endpoints

> * *permissive*

Reviewer: ```suggestion
* *`permissive`*
```

** Addressables should accept events on both HTTP and HTTPS endpoints
** Addressables should advertise both HTTP and HTTPS endpoints
** Producers should prefer sending events to HTTPS endpoints, if available

> * *strict*

Reviewer: ```suggestion
* *`strict`*
```

** Addressables must not accept events to non-HTTPS endpoints
** Addressables must only advertise HTTPS endpoints

== Setup transport encryption in `KnativeEventing`

. Enabling transport-encryption in `KnativeEventing`:
+
[source,yaml]
----
apiVersion: operator.knative.dev/v1beta1
kind: KnativeEventing
metadata:
  name: knative-eventing
  namespace: knative-eventing
spec:

  # Other spec fields omitted ...
  # ...

  config:
    features:
      transport-encryption: strict <1>
----
+

> <1> Configure `transport-encryption` to `strict`.

Reviewer 3: does this only work with `strict`?

Author: no, permissive process is the same, only the config is different

. Apply the `KnativeEventing` resource:
+
[source,terminal]
----
$ oc apply -f <filename>
----

== Configure additional CA bundles

By default, Eventing clients trusts the OpenShift CA bundle that you have configured when using a custom PKI for OpenShift, as documented at https://docs.openshift.com/container-platform/4.14/networking/configuring-a-custom-pki.html.

> If you need to add additional CA bundles for Eventing, you can do so by creating `ConfigMaps` in the `knative-eventing` namespace with label `networking.knative.dev/trust-bundle: true`:

Reviewer 2: Is there nothing that will run outside of eventing namespace that could need the bundle also?

Reviewer 2: @pierDipi just to make sure, you do never deploy an Eventing component into another namespace than `knative-eventing` that would need the bundle?

Author: We deploy some but it's automatically handled for users, probably the only additional point I can add is how SinkBinding injects trust bundles into containers so that users can configure their HTTP clients to trust the CA

Author: Added SB / workloads that connect to Eventing docs

== Adding `SelfSigned` `ClusterIssuer` to CA trust bundles

In case you are using a `SelfSigned` `ClusterIssuer` as described in the <<setup_selfsigned_clusterissuer>> section, you can add the CA to the Eventing CA trust bundles by running the following commands:

. Export the CA from the `knative-eventing-ca` secret in the {certmanageroperatorname} namespace, `cert-manager` by default:
+
[source,terminal]
----
$ oc get secret -n cert-manager knative-eventing-ca -o=jsonpath='{.data.ca\.crt}' | base64 -d > ca.crt
----
. Create a CA trust bundle in the `knative-eventing` namespace:
+
[source,terminal]
----
$ oc create configmap -n knative-eventing my-org-selfsigned-ca-bundle --from-file=ca.crt
----
. Label the `ConfigMap` with `networking.knative.dev/trust-bundle: "true"` label:
+
[source,terminal]
----
$ oc label configmap -n knative-eventing my-org-selfsigned-ca-bundle networking.knative.dev/trust-bundle=true
----

== Ensure seamless CA rotation

Ensuring seamless CA rotation is essential to avoid service downtime, or to deal with an emergency.
The following procedure explains how you can seamlessly rotate a CA.

> 1. Create a CA certificate.

Reviewer 2: ```suggestion
1. Create a new CA certificate.
```

> 2. Add the public key to the CA trust bundles as described in the <<configure_addition_ca_trust_bundles>> section.

Reviewer 2: ```suggestion
2. Add the public key of the new CA to the CA trust bundles as described in the <<configure_addition_ca_trust_bundles>> section. Make sure to also keep the public key of the existing CA.
```

> 3. Ensure that all clients have downloaded the latest set of CA trust bundles.

Reviewer 2: ```suggestion
3. Ensure that all clients have consumed the latest set of CA trust bundles. Knative Eventing components will automatically reload the changed CA bundles. If you have custom workload consuming trust bundles as well, make sure to reload/restart them accordingly.
```

Reviewer 2: Not sure if this is correct. If not, you should probably add that people need to re-rollout components in `knative-eventing`?

4. Update the `knative-eventing-ca-issuer` `ClusterIssuer` to reference the secret containing the CA certificate created at step 1.
5. Force cert-manager to renew certificates in the `knative-eventing` namespace. +

> Refer to the cert-manager documentation for more details: https://cert-manager.io/docs/usage/certificate/#reissuance-triggered-by-user-actions.

Reviewer 2: 6. As soon as the CA rotation is fully completed, you can remove the public key of the old CA from the trust bundle `ConfigMap`
